rcos_io/services/github.py

```diff
@@ -111,3 +112,121 @@ class CommitInfo(TypedDict):
     commit: Commit
     author: User
     committer: User
+
+
+#    html url structure: https://github.com/Apexal/rcos_io
+#     api url structure: https://api.github.com/repos/Apexal/rcos_io
+
+#          get branches: https://api.github.com/repos/Apexal/rcos_io/branches
+#           get commits: https://api.github.com/repos/Apexal/rcos_io/commits
+#   get commits by user: ?author=username or email
+#     get commits since: ?since=ISO-8601 timestamp <-- starttime
+#     get commits until: ?until=ISO-8601 timestamp <-- endtime
+# get commits from head: ?sha=commit sha
+
+
+def gen_params(**kwargs):
+    """
+    Generates params for request querystring.
+    """
+    valid_keys = {"author", "since", "until", "sha"}
+    params = {"per_page": 100}  # get 100 commits per page
+    for arg_key, arg_val in kwargs.items():
+        # only include valid github api parameters
+        # only include non None values
+        if arg_key in valid_keys and arg_val is not None:
+            params[arg_key] = arg_val
+    return params
+
+
+def get_commits(
+    repolink: str,
+    starttime: datetime.datetime,
+    endtime: datetime.datetime = datetime.datetime.now(),
+    user: Optional[str] = None,
+):
+    """
+    Grabs all commits on all branches for a given GitHub repo, after a given time.
+
+    takes >
+         repolink: github html url
+        starttime: datetime object for when to grab commits since
+          endtime: datetime object for when to grab commits until
+             user: username for desired user to filter commits for
+
+    returns >
+        all_commits: dictionary of commits >
+            key: commit -> tree -> sha
+            value: dictionary of commit info >
+                key: "url"
+                value: html url of commit
+                ;
+                key: "timestamp"
+                value: ISO-8601 timestamp of commit
+    """
+
+    # convert datetime objects to iso strings
+    starttime, endtime = starttime.isoformat(), endtime.isoformat()
+
+    # grabs "owner_username/repo_name"
+    repoid = repolink[repolink.index("github.com/") + len("github.com/") :]
```

Comment:
Wouldn't it be better to use something like [urllib.parse](https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse) instead of doing substring-magic here?